# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/ViewportUtils.cpp
# Commit: bda050135c64
# Full Hash: bda050135c64c3e4dc8c69b2be06f495592efc6e
# Author: Botond Ballo <botond@mozilla.com>
# Date: 2020-05-06 09:37:16
# Regressor Bug: 1556556
# File Overlap Count: 1
# Description:
#   Bug 1556556 - Move GetCallbackTransform() into a new ViewportUtils class. r=kats
#   
#   This function (and helper functions that wrap it) will be used extensively
#   throughout layout code, so keeping it in APZCCallbackHelper seems awkward.
#   
# ==============================================================================

diff -r 8206cf323d5b -r bda050135c64 layout/base/ViewportUtils.cpp
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/layout/base/ViewportUtils.cpp	Tue May 05 19:22:12 2020 +0000
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "mozilla/PresShell.h"
+#include "mozilla/ViewportUtils.h"
+#include "mozilla/layers/APZCCallbackHelper.h"
+#include "mozilla/layers/ScrollableLayerGuid.h"
+#include "nsIContent.h"
+#include "nsLayoutUtils.h"
+
+namespace mozilla {
+
+using layers::APZCCallbackHelper;
+using layers::ScrollableLayerGuid;
+
+CSSToCSSMatrix4x4 ViewportUtils::GetCallbackTransform(
+    ScrollableLayerGuid::ViewID aScrollId) {
+  if (aScrollId == ScrollableLayerGuid::NULL_SCROLL_ID) {
+    return {};
+  }
+  nsCOMPtr<nsIContent> content = nsLayoutUtils::FindContentFor(aScrollId);
+  if (!content) {
+    return {};
+  }
+
+  // First, scale inversely by the root content document's pres shell
+  // resolution to cancel the scale-to-resolution transform that the
+  // compositor adds to the layer with the pres shell resolution. The points
+  // sent to Gecko by APZ don't have this transform unapplied (unlike other
+  // compositor-side transforms) because Gecko needs it applied when hit
+  // testing against content that's conceptually outside the resolution,
+  // such as scrollbars.
+  float resolution = 1.0f;
+  if (PresShell* presShell =
+          APZCCallbackHelper::GetRootContentDocumentPresShellForContent(
+              content)) {
+    resolution = presShell->GetResolution();
+  }
+
+  // Now apply the callback-transform. This is only approximately correct,
+  // see the comment on GetCumulativeApzCallbackTransform for details.
+  CSSPoint transform = nsLayoutUtils::GetCumulativeApzCallbackTransform(
+      content->GetPrimaryFrame());
+
+  return mozilla::CSSToCSSMatrix4x4::Scaling(1 / resolution, 1 / resolution, 1)
+      .PostTranslate(transform.x, transform.y, 0);
+}
+
+}  // namespace mozilla
\ No newline at end of file